easybuild/easyblocks/b/boost.py

```diff
@@ -224,57 +223,80 @@ def build_step(self):
             else:
                 cxxflags += '0'
         if cxxflags is not None:
-            bjamoptions += " cxxflags='%s'" % cxxflags
+            self.bjamoptions += " cxxflags='%s'" % cxxflags
         ldflags = os.getenv('LDFLAGS')
         if ldflags is not None:
-            bjamoptions += " linkflags='%s'" % ldflags
+            self.bjamoptions += " linkflags='%s'" % ldflags
 
         # specify path for bzip2/zlib if module is loaded
         for lib in ["bzip2", "zlib"]:
             libroot = get_software_root(lib)
             if libroot:
-                bjamoptions += " -s%s_INCLUDE=%s/include" % (lib.upper(), libroot)
-                bjamoptions += " -s%s_LIBPATH=%s/lib" % (lib.upper(), libroot)
+                self.bjamoptions += " -s%s_INCLUDE=%s/include" % (lib.upper(), libroot)
+                self.bjamoptions += " -s%s_LIBPATH=%s/lib" % (lib.upper(), libroot)
 
-        paracmd = ''
+        self.paracmd = ''
         if self.cfg['parallel']:
-            paracmd = "-j %s" % self.cfg['parallel']
+            self.paracmd = "-j %s" % self.cfg['parallel']
 
         if self.cfg['only_python_bindings']:
             # magic incantation to only install Boost Python bindings is... --with-python
             # see http://boostorg.github.io/python/doc/html/building/installing_boost_python_on_your_.html
-            bjamoptions += " --with-python"
+            self.bjamoptions += " --with-python"
+
+        self.log.info("Building boost with single and multi threading")
+        self.bjamoptions += " threading=single,multi --layout=tagged"
 
         if self.cfg['boost_mpi']:
             self.log.info("Building boost_mpi library")
-            self.build_boost_variant(bjamoptions + " --with-mpi", paracmd)
+            mpi_bjamoptions = " --with-mpi"
+            self.build_boost_variant(self.bjamoptions + mpi_bjamoptions, self.paracmd)
```

Comment:
The goal should be to do a single invocation of b2 with everything. To use mpi you can add `using mpi ;` to the user-config.jam file. I'll double check if that is the same

Reply:
Yeah, didn't get that to work with my first tries, so I decided to PR it as it is to start with...

Reply:
I'm not so sure about that. Think Boost.MPI the same way we have Boost.Python, then we would in reality only want it to build the mpi parts, i.e. using --with-mpi, but it would then be the only build/install step it does, just like for the python part.

Reply:
Yes, but that is not what it currently is: `boost_mpi = True` in the EC additionally installs Boost.MPI and we can't change that until EB 5.x
However it makes sense providing a (new) knob to build only Boost.MPI (same as Boost.Python) because we moved Boost to GCCcore, i.e. no MPI there.
We might even have `boost_mpi = None` being the default and the EasyBlock decides whether to build with or without MPI based on the currently loaded modules/dependencies.

Reply:
We could simply check if name = Boost.MPI do --with-mpi and only that (and verify that usempi = True first)
Else if either boost_mpi=True (requires usempi=True) or just usempi = True then add "using mpi" to config and build everything, perhaps verifying that we have MPI in the toolchain in all cases.

usempi=True/False should be the determining factor for MPI since that's what is mostly used in other easyblocks and for simpler easyconfigs.

Reply:
But all this should probably go in a separate PR...

Reply:
>But all this should probably go in a separate PR...

Yes. But in this PR we should keep support for `boost_mpi` as it was before, i.e. build it **additionally**, hence add the `using mpi`

Reply:
Nah, I'd prefer not to change this any more, it does the "same" thing as the old one, i.e. multiple b2 invocations when boost_mpi=True, producing all the libs.
It's better to deal with this MPI stuff in a separate PR.

I.e. my Boost-1.74.0-gompi-2020b.eb which has boost_mpi=True produces all the libs of standard boost + the mpi ones and creates the correct links when using this boost.py.

Reply:
Hm, but multiple invocations of b2 was the source of the problem that lead to this. So IMO we should fix that together. I can offer to tackle the remaining issues on top of this. But yeah, maybe in a separate PR, I guess doesn't matter much

Reply:
No multiple invocations was not really the problem, it was doing it with layout=system once and layout=tagged once. Now I always do layout=tagged so there is no problem with multiple invocations.

Reply:
I mean: if we do a single invocation, then there is no room for such mistakes at all.

Reply:
I agree that this can be changed in a follow up PR, let's get this one merged as is to avoid dragging it along...